src/Samples/Sarif.WorkItems.Sample/Program.cs

```diff
@@ -5,112 +5,83 @@
+            SarifLog matchedSarif = resultMatcher.CreateBaselinedSarifLog(
+                baselineSarif, 
+                currentSarif);
 
-            // 3. Make sure everything has a GUID and that we have a map back to each result from it
-            var guidToMatchedResultMap = new Dictionary<string, Result>();
-            foreach (Result result in matchedLog.Runs[0].Results)
-            {
-                if (result.Guid == null)
-                {
-                    result.Guid = Guid.NewGuid().ToString();
-                }
-                guidToMatchedResultMap[result.Guid] = result;
-            }
+            // 3. We need a map from each result guid to the matched SARIF result.
+            //    As bugs are filed against refactored files, we use the guid to 
+            //    link to the original file in order to populate the WIT URI.
 
-            var adoClient = new AzureDevOpsClient();
-            var filingTarget = new AzureDevOpsFilingTarget(adoClient);
+            var guidMappingVisitor = new GuidMappingVisitor();
+            guidMappingVisitor.Visit(matchedSarif);
 
-            var workItemFiler = new WorkItemFiler(
-                filingTarget,
-                filteringStrategy: new NewResultsFilteringStrategy(),
-                groupingStrategy: new OneResultPerWorkItemGroupingStrategy());
+            // 4. Now we need to refactor the log file into one or more log files, 
+            // each of which will drive the creation of a single bug.
 
-            string logFilePath = Path.GetTempFileName();
-            File.WriteAllText(logFilePath, JsonConvert.SerializeObject(matchedLog));
+            IEnumerable<SarifLog> sarifLogsToFile = SarifSplitter.Split(
+                sarifLog: matchedSarif,
+                filterPredicate: (result) => { return result.BaselineState == BaselineState.New; },
+                splittingStrategy: SplittingStrategy.OneFilePerRunPerRule);
 
-            // IMPORTANT! There's no facility defined yet for creating/configuring titles, area paths, etc.
+            // 5. Now we have a set of SARIF files, each of which will drive creation of a single bug
+            IEnumerable<WorkItemFilingMetadata> workItemsToFile = SarifLog.CreateWorkItemFilingMetadataFromLogs(sarifLogsToFile);
 
-            try
-            {
-                IEnumerable<ResultGroup> resultGroups = workItemFiler.FileWorkItems(logFilePath).Result;
-                foreach (ResultGroup resultGroup in resultGroups)
-                {
-                    foreach (Result result in resultGroup.Results)
-                    {
-                        bool matched = guidToMatchedResultMap.TryGetValue(result.Guid, out Result originalResult);
-                        Debug.Assert(matched);
-                        originalResult.WorkItemUris = result.WorkItemUris;
-                    }
-                }
-            }
-            finally
+            AdoWorkItemFiler adoWorkItemFiler = new AdoWorkItemFiler()
```

Comment:
>AdoWorkItemFiler [](start = 52, length = 16)

I presume that this and a `GitHubWorkItemFiler` would derive from an ABC `WorkItemFiler`. Why did you choose this approach rather than injecting a `WorkItemHost` abstraction (which is my current design)?

End users would not have to do they injection. We would provide a factory:
```
public static class WorkItemFilerFactory
{
    public static WorkItemFiler CreateAdoWorkItemFiler() { return new WorkItemFiler(new AdoWorkItemHost()); }
    public static WorkItemFiler CreateGitHubWorkItemFiler() { return new WorkItemFiler(new GitHubWorkItemHost(); }
}

Now the caller does this:
```
WorkItemFiler filer = WorkItemFilerFactory.CreateAdoWorkItemFiler();
```
Callers only have to worry about one class instead of one per host type.

Reply:
This all assumes there will be no difference in helpers provided for an ADO-specific work item filer and I predict this is not the case. And so (new AdoWorkItemFiler()) that provides some core functionality and additional content is very likely. ie. basic object inheritance.  Glad to discuss further. This sort of API evolution is simple to absorb (i.e., breaking a core type into more granular derived types), so I don't see it as an urgent topic.

---
In reply to: [307087340](https://github.com/microsoft/sarif-sdk/pull/1589#discussion_r307087340) [](ancestors = 307087340)